oeps/oep-0008.rst

```diff
@@ -0,0 +1,323 @@
+
+
+Task Management
+---------------
+We use Makefiles for task management. The reason for this choice is that
+almost any system no matter the OS should be able to use them without any
+special software installation or configuration.
+
+If the Makefile gets too complex, use it as a wrapper to `Paver`_ scripts,
+which are written in Python.
+
+Standard Makefile Targets:
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
++--------------+--------------------------------------------------------------+
+| target       | description                                                  |
++==============+==============================================================+
+| help         | display a help message                                       |
++--------------+--------------------------------------------------------------+
+| pip-compile  | update the ``requirements/*.txt`` files with the             |
+|              | latest packages satisfying ``requirements/*.in``             |
++--------------+--------------------------------------------------------------+
+| requirements | install development environment requirements                 |
++--------------+--------------------------------------------------------------+
+| quality      | check coding style                                           |
++--------------+--------------------------------------------------------------+
+| test         | run all unit, integration, and acceptance tests for the repo |
++--------------+--------------------------------------------------------------+
+| validate     | run tests and quality checks                                 |
++--------------+--------------------------------------------------------------+
+| e2e          | run End to End tests against a deployed environment to       |
+|              | verify the system as a user, initiating their actions        |
+|              | through the UI                                               |
++--------------+--------------------------------------------------------------+
+
+.. _Travis CI: https://travis-ci.org/
+.. _PyPI: https://pypi.python.org/pypi
+.. _Paver: https://pythonhosted.org/Paver/
+
+
+Python Language Specific
+------------------------
+
+Requirements
+^^^^^^^^^^^^
+
+* Core requirements are listed in the ``install_requires`` section of
+  ``setup.py``, and only modified when code changes are made that require a
+  new package or drop support for a previously supported one.
+
+* Requirements are managed via `pip-tools`_ and organized by which ones need
+  to be installed for different Python virtual environments.
+
+* Requirements files are kept either at root level or within a directory named
+  ``requirements``.
+
+* Direct dependencies are listed in ``requirements/*.in`` for use in
+  development and `Tox`_ virtual environments with as few version constraints
+  as possible.
+
+* Full pinned dependencies in ``requirements/*.txt`` are auto-generated via
+  ``make pip-compile``.
```

Comment:
We don't have much experience yet with pip-tools. I don't understand how privately installed packages (like pudb) will be managed while pip-tools tries to uninstall packages that shouldn't be there.  I also don't understand how we will manage pinning if pip-tools wants to do an all-or-nothing upgrade to the latest versions.  That will be changing many versions all at once.  Is that what we want?